src/main/java/ch/njol/skript/SkriptAddon.java

```diff
@@ -21,123 +21,195 @@
+	 *
+	 * @param withClass A consumer that will run with each found class.
+	 * @param initialize Whether classes found in the package search should be initialized.
+	 * @param basePackage The base package to start searching in (e.g. 'ch.njol.skript').
+	 * @param recursive Whether to recursively search through the subpackages provided.
+	 * @param subPackages Specific subpackages to search in (e.g. 'conditions')
+	 *                    If no subpackages are provided, all subpackages of the base package will be searched.
 	 * @return This SkriptAddon
 	 */
-	public SkriptAddon loadClasses(String basePackage, String... subPackages) throws IOException {
-		assert subPackages != null;
-		JarFile jar = new JarFile(getFile());
+	@SuppressWarnings("ThrowableNotThrown")
+	public SkriptAddon loadClasses(@Nullable Consumer<Class<?>> withClass, boolean initialize, String basePackage, boolean recursive, String... subPackages) {
 		for (int i = 0; i < subPackages.length; i++)
 			subPackages[i] = subPackages[i].replace('.', '/') + "/";
 		basePackage = basePackage.replace('.', '/') + "/";
-		try {
-			List<String> classNames = new ArrayList<>();
 
-			for (JarEntry e : new EnumerationIterable<>(jar.entries())) {
-				if (e.getName().startsWith(basePackage) && e.getName().endsWith(".class")) {
+		int depth = !recursive ? StringUtils.count(basePackage, '/') + 1 : 0;
+
+		File file = getFile();
+		if (file == null) {
+			Skript.error("Unable to retrieve file from addon '" + getName() + "'. Classes will not be loaded.");
+			return this;
+		}
+
+		try (JarFile jar = new JarFile(file)) {
+			List<String> classNames = new ArrayList<>();
+			boolean hasWithClass = withClass != null;
+			if (entryCache == null)
+				entryCache = jar.stream().toArray(JarEntry[]::new);
+			for (int i = 0; i < entryCache.length; i++) {
+				JarEntry e = entryCache[i];
+				if (e == null) // This entry has already been loaded before
+					continue;
+				String name = e.getName();
+				if (name.startsWith(basePackage) && name.endsWith(".class") && (recursive || StringUtils.count(name, '/') <= depth)) {
 					boolean load = subPackages.length == 0;
-					for (String sub : subPackages) {
-						if (e.getName().startsWith(sub, basePackage.length())) {
+					for (String subPackage : subPackages) {
+						if (e.getName().startsWith(subPackage, basePackage.length())) {
 							load = true;
 							break;
 						}
 					}
-
-					if (load)
+					if (load) {
 						classNames.add(e.getName().replace('/', '.').substring(0, e.getName().length() - ".class".length()));
+						entryCache[i] = null; // Remove this item from the entry cache as this method will only load it once
+					}
 				}
 			}
-
 			classNames.sort(String::compareToIgnoreCase);
-
 			for (String c : classNames) {
 				try {
-					Class.forName(c, true, plugin.getClass().getClassLoader());
+					Class<?> clazz = Class.forName(c, initialize, plugin.getClass().getClassLoader());
+					if (hasWithClass)
```

Comment:
```suggestion
					if (withClass != null)
```
Variable only used once, can also be removed above